File: medium/Search2DMatrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        i, j = 0, len(matrix)-1
        l, r = 0, len(matrix[0])-1

        while i <= j:
            curr_row = i + ((j-i)//2)
            if matrix[curr_row][0] > target:
                j = curr_row-1
            elif matrix[curr_row][-1] < target:
                i = curr_row+1
            else:
                break

        while l <= r:
            curr = l + ((r-l)//2)
            if matrix[curr_row][curr] == target:
                return True
            elif matrix[curr_row][curr] > target:
                r = curr - 1
            elif matrix[curr_row][curr] < target:
                l = curr + 1 
        return False

File: medium/Search2DMatrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // find row
        // if in range binary search
        int top = 0;
        int l = 0;
        int r = matrix[0].size()-1;
        int bot = matrix.size()-1;

        while(top <= bot){
            int curr_row = (top+bot)/2;
            if(matrix[curr_row][0] > target){
                bot = curr_row-1;
            }
            else if(matrix[curr_row][r] < target){
                top = curr_row+1;
            }
            else{
                break;
            }
        }

        int c_row = (top+bot)/2;
        if (c_row < 0 || c_row >= matrix.size()){
            return false;
        }
        while(l <= r){
            int curr_col = (l+r)/2;
            if(matrix[c_row][curr_col] == target){
                return true;
            }
            else if(matrix[c_row][curr_col] >= target){
                r = curr_col-1;
            }
            else{
                l = curr_col+1;
            }
        }
        return false;
    }
};
